### ISSUE TITLE
How to write Info for typedef?

### ISSUE THREAD

**Issue by zjulzq** [bytedeco/javacpp-presets#192]
In my case, there are many typedefs as below:

```
typedef char TThostFtdcTraderIDType[21];
typedef char TThostFtdcInvestorIDType[13];
typedef char TThostFtdcBrokerIDType[11];
```

I have tried these but I failed:
```
infoMap.put(new Info("TThostFtdcTraderIDType").cast().valueTypes("byte[]"));
infoMap.put(new Info("TThostFtdcInvestorIDType").cast().valueTypes("byte[]"));
infoMap.put(new Info("TThostFtdcBrokerIDType").cast().valueTypes("byte[]"));
```

I am in trouble now. Please help me.

**Comment by saudet** [comment]
Using "pointerTypes" instead of "valueTypes" should work ...

**Comment by zjulzq** [comment]
Still meet errors:

<pre>
/home/lzq/workspace/javacpp-presets/ctp/target/classes/org/bytedeco/javacpp/libthostmduerapi.so.cpp: In function ‘_jobject* Java_org_bytedeco_javacpp_ThostFtdcMdApi_00024CThostFtdcReqUserLoginField_TradingDay___3B(JNIEnv*, jobject, jbyteArray)’:
/home/lzq/workspace/javacpp-presets/ctp/target/classes/org/bytedeco/javacpp/libthostmduerapi.so.cpp:101144:21: error: invalid array assignment
     ptr->TradingDay = *(TThostFtdcDateType*)ptr0;
                     ^
/home/lzq/workspace/javacpp-presets/ctp/target/classes/org/bytedeco/javacpp/libthostmduerapi.so.cpp: In function ‘_jobject* Java_org_bytedeco_javacpp_ThostFtdcMdApi_00024CThostFtdcReqUserLoginField_UserProductInfo___3B(JNIEnv*, jobject, jbyteArray)’:
/home/lzq/workspace/javacpp-presets/ctp/target/classes/org/bytedeco/javacpp/libthostmduerapi.so.cpp:101176:26: error: invalid array assignment
     ptr->UserProductInfo = *(TThostFtdcProductInfoType*)ptr0;
</pre>

In my case, I download the API files from a website. There are 2 so files and 4 header files.

1. libthostmduserapi.so
1. libthosttraderapi.so
1. ThostFtdcMdApi.h
1. ThostFtdcTraderApi.h
1. ThostFtdcUserApiDataType.h
1. ThostFtdcUserApiStruct.h

There are actually 2 apis; mduser api and trader api.

1. For mduser api, 4 files are required: `libthostmduserapi.so, ThostFtdcMdApi.h, ThostFtdcUserApiDataType.h, ThostFtdcUserApiStruct.h`.
1. For trader api, 4 files are required: `libthosttraderapi.so, ThostFtdcTraderApi.h, ThostFtdcUserApiDataType.h, ThostFtdcUserApiStruct.h`.
1. The mduser and trader api shares 2 header files: `ThostFtdcUserApiDataType.h, ThostFtdcUserApiStruct.h`.

## Is my solution ok?

1. put the 2 so files in cppbuild/linux-x86_64/lib.
1. put the 4 header files in cppbuild/linux-x86_64/include.
1. build with cmd: `mvn clean install -Djavacpp.platform=linux-x86_64 -Djavacpp.platform.dependency=false -Djavacpp.cppbuild.skip=true --projects .,ctp`

**Comment by zjulzq** [comment]
## Is the @Properties configuration ok?

My configuration in java files:

```
@Properties(target = "org.bytedeco.javacpp.ThostFtdcTraderApi", value = { @Platform(include = { "ThostFtdcTraderApi.h",
		"ThostFtdcUserApiDataType.h", "ThostFtdcUserApiStruct.h" }, library = "libthosttraderapi.so") })
public class ThostFtdcTraderApi implements InfoMapper {
	public void map(InfoMap infoMap) {
		infoMap.put(new Info("TRADER_API_EXPORT").skip());
		infoMap.put(new Info("THOST_TE_RESUME_TYPE").cast().valueTypes("int"));
		infoMap.put(new Info("TThostFtdcTraderIDType").cast().pointerTypes("byte[]"));
		infoMap.put(new Info("TThostFtdcInvestorIDType").cast().pointerTypes("byte[]"));
		infoMap.put(new Info("TThostFtdcBrokerIDType").cast().pointerTypes("byte[]"));
		......
	}
```

```
@Properties(target = "org.bytedeco.javacpp.ThostFtdcMdApi", value = { @Platform(include = { "ThostFtdcMdApi.h",
		"ThostFtdcUserApiDataType.h", "ThostFtdcUserApiStruct.h" }, library = "libthostmduerapi.so") })
public class ThostFtdcMdApi implements InfoMapper {
	public void map(InfoMap infoMap) {
		infoMap.put(new Info("MD_API_EXPORT").skip());
		infoMap.put(new Info("THOST_TE_RESUME_TYPE").cast().valueTypes("int"));
		infoMap.put(new Info("TThostFtdcTraderIDType").cast().pointerTypes("byte[]"));
		infoMap.put(new Info("TThostFtdcInvestorIDType").cast().pointerTypes("byte[]"));
		infoMap.put(new Info("TThostFtdcBrokerIDType").cast().pointerTypes("byte[]"));
		infoMap.put(new Info("TThostFtdcBrokerAbbrType").cast().pointerTypes("byte[]"));
		......
	}
```

**Comment by zjulzq** [comment]
@saudet I am waiting for your help.

**Comment by saudet** [comment]
"library" should be "link", but yes other than that, it's fine.

Arrays are a tricky type in C... I'm not sure I've added what's needed to handle that case properly.

**Comment by zjulzq** [comment]
@saudet I fix the `library` to `link`. The `invalid array assignment` problem exists.

## Is there a easy way to fix it?

I want to apply javacpp to simplify my work. I used to convert C++ API to C API, then use JNA to access it. That is a terrible solution.

**Comment by saudet** [comment]
I've made some changes, so that we can now do something like this:
```java
.put(new Info("TThostFtdcTraderIDType", "TThostFtdcInvestorIDType", "TThostFtdcBrokerIDType", "TThostFtdcBrokerAbbrType", ... )
        .cast().cppTypes("const char*").valueTypes("BytePointer"))
```
Let me know if that doesn't work for you! Thanks for reporting

**Comment by zjulzq** [comment]
It works for me. Thank you.

**Comment by saudet** [comment]
Fix included the latest version of 1.2! Thanks for reporting
